### ISSUE TITLE
question: Why tabby use less ram than wasmedge?

### ISSUE THREAD

**Issue by katopz** [second-state/llama-utils#18]
### Summary

i found something interesting about tabby vs wasmedge RAM usage on m3max.
![image](https://github.com/second-state/llama-utils/assets/97060/0b1fd0ec-b6a9-4b87-98b2-de2253791feb)

> https://github.com/TabbyML/tabby
```
tabby serve --device metal --model TabbyML/DeepseekCoder-6.7B --chat-model TabbyML/Mistral-7B
```
vs
```
wasmedge-ggml-llama-interactive % wasmedge --dir .:. --nn-preload default:GGML:AUTO:mistral-7b-instruct-v0.1.Q5_K_M.gguf llama-api-server.wasm -p chatml -r '<|im_end|>' -s 0.0.0.0:8081

curl -X POST http://localhost:8081/v1/chat/completions -H 'accept:application/json' -H 'Content-Type: application/json' -d '{"messages":[{"role":"system", "content": "You are a helpful assistant."}, {"role":"user", "content": "Write helloworld code in Rust"}], "model":"MistralLite-7B"}'
```
TLDR; tabby use 2x less RAM and for 2 models

### Appendix

Not sure why and maybe we can improve somewhere to achieve the same ram usage?

**Comment by hydai** [comment]
Hi @katopz 
Are both models the same? With the same quantize and size?

**Comment by katopz** [comment]
Not exactly same

- tabby: `Mistral-7B.q8_0.v2.gguf` + `DeepseekCoder-6.7B.q8_0.v2.gguf` = 7.7GB + 7.16GB
- wasmedge: `mistral-7b-instruct-v0.1.Q5_K_M.gguf` = 5.13GB

And tabby get even better for both `Mistral-7B` it use only 1.08GB 🤨

```
tabby serve --device metal --model TabbyML/Mistral-7B --chat-model TabbyML/Mistral-7B
```

<img width="740" alt="image" src="https://github.com/second-state/llama-utils/assets/97060/95997e2d-2e91-4164-a4cf-3d24f42beea3">

**Comment by hydai** [comment]
I am not sure how they handle the model. Will Tabby load a whole model into memory?
The API server will load the model into memory until the server is closed, so memory usage is expected to be close to the model size.

**Comment by katopz** [comment]
look like it did like as usual via https://github.com/TabbyML/llama.cpp binding

> https://github.com/TabbyML/tabby/blob/main/crates/tabby/src/services/model.rs
```rust
fn create_ggml_engine(device: &Device, model_path: &str, parallelism: u8) -> impl TextGeneration {
    let options = llama_cpp_bindings::LlamaTextGenerationOptionsBuilder::default()
        .model_path(model_path.to_owned())
        .use_gpu(device.ggml_use_gpu())
        .parallelism(parallelism)
        .build()
        .unwrap();

    llama_cpp_bindings::LlamaTextGeneration::new(options)
}
```
